parts/k8s/cloud-init/artifacts/cse_config.sh

```diff
@@ -282,6 +282,9 @@ installContainerd() {
 ensureContainerd() {
   wait_for_file 1200 1 /etc/systemd/system/containerd.service.d/exec_start.conf || exit {{GetCSEErrorCode "ERR_FILE_WATCH_TIMEOUT"}}
   wait_for_file 1200 1 /etc/containerd/config.toml || exit {{GetCSEErrorCode "ERR_FILE_WATCH_TIMEOUT"}}
+  {{- if HasContainerDataDir}}
+  echo -e "root = \"{{GetContainerDataDir}}\"\n$(cat /etc/containerd/config.toml)" > /etc/containerd/config.toml
```

Comment:
So, this is ugly. An alternative is doing it in cloud-init for non-vhd installs and doing it in the VHD otherwise. Thoughts?

Reply:
So, we don't want to just get the whole generic key/val implementation over with first of all? See the way we use the `GetKubeletConfigKeyVals` go template layer func in cloud-init (`parts/k8s/cloud-init/masternodecustomdata.yml` and `parts/k8s/cloud-init/nodecustomdata.yml`) to accommodate generic key/val injection at runtime.

Reply:
The part I wasn't sure about was how much to generalize this for docker vs containerd because the config files are totally different.

Are you thinking we could do a func like for kubelet, 1 each for docker and containerd, and have those functions accept the raw `map[string]string` and spit out the raw CRI-specific JSON? Then we can avoid any conditionals in the bash

Reply:
Exactly 💯 

Not trying to introduce scope creep, but I think we'll thank ourselves later for doing it the generic way. Thoughts?

Reply:
I'd really like to avoid "containerdConfig" and "dockerConfig", but we're kind of hiding it no matter what

Reply:
i'm on board with that, will update

Reply:
I think we can stick with a single `containerRuntimeConfig`, and then have per-runtime go template helper funcs. As long as we can assume that key/val strings are common across all runtimes (I think we can), then we can use a single interface.

Reply:
is there a good way to handle errors in the template funcs? I want to do something like this, but struggling to find a good place to drop the Marshal to catch the error. I could do raw string manipulation, but that feels error prone (pun intended) and I wouldn't be guaranteed to construct valid json anyway.

```go
// GetDockerConfig returns the full docker daemon configuration including user overrides.
func (k *KubernetesConfig) GetDockerConfig() (string, error) {
	config := map[string]interface{}{
		"live-restore": true,
		"log-driver": "json-file",
		"log-opts": {
			"max-size": "50m",
			"max-file": "5",
		},
	}

	dataDir, ok := k.ContainerRuntimeConfig[ContainerDataDirKey]
	if ok {
		config["data-root"] = dataDir
	}

	b, err := json.MarshalIndent(config, "", "    ")
	return string(b), err
}
```

Reply:
and containerd uses toml, even more fun